operators/go/submariner_controller.go

```diff
@@ -0,0 +1,221 @@
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileSubmariner{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("submariner-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Submariner
+	err = c.Watch(&source.Kind{Type: &submarinerv1alpha1.Submariner{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource Pods and requeue the owner Submariner
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &submarinerv1alpha1.Submariner{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileSubmariner implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileSubmariner{}
+
+// ReconcileSubmariner reconciles a Submariner object
+type ReconcileSubmariner struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Submariner object and makes changes based on the state read
+// and what is in the Submariner.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileSubmariner) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Submariner")
+
+	// Fetch the Submariner instance
+	instance := &submarinerv1alpha1.Submariner{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+  // TODO: Add deployment-creation logic here? See memc example
+
+  // Create submariner-engine SA
+  //subm_engine_sa := corev1.ServiceAccount{}
+  //subm_engine_sa.Name = "submariner-engine"
+  //reqLogger.Info("Created a new SA", "SA.Name", subm_engine_sa.Name)
+
+  // FIXME: CRDs need to be created before the Operator starts. Use OLM.
```

Comment:
If need to be created by the OLM, may be just add the comment here that CRDs must be created by OLM itself and remove 104-138 ?

Reply:
Yeah, agreed, I will remove this commented-out code. Was just keeping it around as a nice example of what we needed, but it's more confusing than helpful at this point.

Reply:
This is done in the latest version.